fortanix-vme/nitro-attestation-verify/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+pub struct Unverified;
+impl VerificationType for Unverified{}
+
+pub trait SafeToDeserialize {}
+impl SafeToDeserialize for Unverified{}
+
+#[derive(Debug, Deserialize)]
+#[serde(bound(deserialize = "V: SafeToDeserialize"))]
+pub struct AttestationDocument<V: VerificationType = Verified> {
+    module_id: String,
+    timestamp: u64,
+    digest: String,
+    pcrs: BTreeMap<u32, ByteBuf>,
+    certificate: ByteBuf,
+    cabundle: Vec<ByteBuf>,
+    public_key: Option<ByteBuf>,
+    user_data: Option<ByteBuf>,
+    nonce: Option<ByteBuf>,
+    #[serde(skip)]
+    cose: Option<CoseSign1>,
+    #[serde(skip)]
+    type_: PhantomData<V>,
+}
+
+impl<V: VerificationType> AttestationDocument<V> {
+    pub fn module_id(&self) -> &String {
+        &self.module_id
+    }
+
+    pub fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+
+    pub fn digest(&self) -> &String {
+        &self.digest
+    }
+
+    pub fn pcrs(&self) -> &BTreeMap<u32, ByteBuf> {
+        &self.pcrs
+    }
+
+    pub fn certificate(&self) -> &ByteBuf {
+        &self.certificate
+    }
+
+    pub fn cabundle(&self) -> &Vec<ByteBuf> {
+        &self.cabundle
+    }
+
+    pub fn public_key(&self) -> Option<&ByteBuf> {
+        self.public_key.as_ref()
+    }
+
+    pub fn user_data(&self) -> Option<&ByteBuf> {
+        self.user_data.as_ref()
+    }
+
+    pub fn nonce(&self) -> Option<&ByteBuf> {
+        self.nonce.as_ref()
+    }
+}
+
+impl AttestationDocument<Unverified> {
+    pub fn from_slice(data: &[u8]) -> Result<Self, NitroError> {
+        // https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html
+        let cose = CoseSign1::from_bytes(data)
+                    .map_err(|err| NitroError::CoseParsingFailure(format!("cose error: {:?}", err)))?;
+        let payload = cose.get_payload::<Mbedtls>(None)
+                        .map_err(|err| NitroError::CoseParsingFailure(format!("cose error: {:?}", err)))?;
+        let mut doc: AttestationDocument<Unverified> = serde_cbor::from_slice(&payload)
+                                                        .map_err(|e| NitroError::PayloadParsingFailure(format!("payload error: {}", e)))?;
+        doc.cose = Some(cose);
+        Ok(doc)
+    }
+
+    /// Verifies the `AttestationDocument` given `root_certs` as DER-encoded trusted root CAs
+    pub fn verify(self, root_certs: &[Vec<u8>]) -> Result<AttestationDocument, NitroError> {
+        self.verify_ex(root_certs, verify_certificates)
+    }
+
+    /// Expert version of the `verify` function. Verifies the `AttestationDocument` given `root_certs` as
+    /// DER-encoded trusted root CAs and a `verify_certs` function that takes the signing
+    /// certificate, CA bundle and root certificates as input.
+    pub fn verify_ex<F>(self, root_certs_der: &[Vec<u8>], verify_certs: F) -> Result<AttestationDocument, NitroError>
+        where
+        F: FnOnce(&ByteBuf, &Vec<ByteBuf>, &[Vec<u8>]) -> Result<MbedtlsBox<Certificate>, NitroError>
```

Comment:
what's the use case for this generic function?

Since generic returns MbedtlsBox we can't use this abstraction to migrate to openssl if anyone wants to.

Reply:
Expired certificates may still be accepted when running in CI. The use of `faketime` may not be that easy in every situation. So I created a way to avoid the expiration check, as you requested. :)